src/DataStructures/Tensor/TensorData.hpp

```diff
@@ -45,49 +50,46 @@ bool operator!=(const TensorComponent& lhs, const TensorComponent& rhs);
- * grid points in the given dimension. The `TensorComponent`s must live on the
- * grid of the size of the extents. We use runtime extents instead of the
- * `Index` class because observers may write 1D, 2D, or 3D data in a 3D
- * simulation.
+ * \brief Holds tensor components on a grid, to be written into an H5 file
  */
-struct ExtentsAndTensorVolumeData {
-  ExtentsAndTensorVolumeData() = default;
-  ExtentsAndTensorVolumeData(std::vector<size_t> extents_in,
-                             std::vector<TensorComponent> components);
-
-  // NOLINTNEXTLINE(google-runtime-references)
-  void pup(PUP::er& p);
-  std::vector<size_t> extents{};
-  std::vector<TensorComponent> tensor_components{};
-};
-
-bool operator==(const ExtentsAndTensorVolumeData& lhs,
-                const ExtentsAndTensorVolumeData& rhs);
-bool operator!=(const ExtentsAndTensorVolumeData& lhs,
-                const ExtentsAndTensorVolumeData& rhs);
-
-/*!
- * An extension of `ExtentsAndTensorVolumeData` to store `Spectral::Quadrature`
- * and `Spectral::Basis`  associated with each axis of the element, in addition
- * to the extents and tensor components data.
- */
-struct ElementVolumeData : ExtentsAndTensorVolumeData {
+struct ElementVolumeData {
   ElementVolumeData() = default;
-  ElementVolumeData(std::vector<size_t> extents_in,
+  ElementVolumeData(std::string element_name_in,
                     std::vector<TensorComponent> components,
+                    std::vector<size_t> extents_in,
                     std::vector<Spectral::Basis> basis_in,
-                    std::vector<Spectral::Quadrature> quadrature_in,
-                    std::string element_name_in);
+                    std::vector<Spectral::Quadrature> quadrature_in);
+  template <size_t Dim>
+  ElementVolumeData(const ElementId<Dim>& element_id,
+                    std::vector<TensorComponent> components,
+                    const Mesh<Dim>& mesh);
 
   // NOLINTNEXTLINE(google-runtime-references)
   void pup(PUP::er& p);
+  /// Name of the grid (should be human-readable). For standard volume data this
+  /// name must be the string representation of an ElementId, such as
+  /// [B0,(I1L0,I0L0,I2L3)]. Code that reads the volume data may rely on this
+  /// pattern to reconstruct the ElementId.
+  std::string element_name{};
+  /// All tensor components on the grid
+  std::vector<TensorComponent> tensor_components{};
+  /// Number of grid points in every dimension of the grid
+  std::vector<size_t> extents{};
+  /// Spectral::Basis in every dimension of the grid
   std::vector<Spectral::Basis> basis{};
+  /// Spectral::Quadrature in every dimension of the grid
   std::vector<Spectral::Quadrature> quadrature{};
-  std::string element_name{};
 };
 
 bool operator==(const ElementVolumeData& lhs, const ElementVolumeData& rhs);
 bool operator!=(const ElementVolumeData& lhs, const ElementVolumeData& rhs);
+
+template <size_t Dim>
+ElementVolumeData::ElementVolumeData(const ElementId<Dim>& element_id,
```

Comment:
Why not put this in the cpp file along with the other constructor?

Reply:
We can't have DataStructures depend on Domain (where ElementId is), so I just forward declare here. I guess we could consider moving this TensorData file around, but it's not obvious to me where exactly, so I think this is an ok solution for now.

Reply:
Ahh I see. This is ok then.

There doesn't seem to be anything super special about `ElementVolumeData` that it has to be in DataStructures. `TensorComponent` I think should stay. But maybe `ElementVolumeData` can be moved somewhere else so it can depend on Domain? However, that's beyond the scope of this PR though.

Reply:
Yea, DataStructures can't depend on Domain. However, adding this dependency in a header file already creates that dependency, right?

Reply:
The forward declaration is fine, dependency-wise. The calling code has to a include DomainStructure/ElementId.hpp, which is also fine I think.

Reply:
That's still a dependency, though. You need to have `INTERFACE Domain` in CMake. It won't break because of cyclic dependencies at the shared lib level, but in general I think we've been avoiding that and the cyclic dependency detection CMake code I wrote counts those as "bad"

Reply:
True. Technically I don't have to add the `INTERFACE Domain` in CMake to get things to compile (though it would be the correct thing to do). I don't see a great solution because I want to make sure that get_output(element_id) is used for grid names at a very low level so we can rely on it when reading volume files back in. Also I want to get extents/basis/quadrature from a Mesh because it's just annoying to deal with otherwise. All I can think of is put a factory method into `Observer`. It's not as low-level as I want it to be, but a compromise?

Reply:
Actually we could depend H5 on DomainStructure (not Domain). Thoughts?

Reply:
If you can get that to work, I think it makes sense. Our volume H5 files have to know things about the domain structure regardless, so it follows that the library would depend on DomainStructure.

Reply:
If `Domain` depends on H5 then that won't work. I'm not sure that's the case, though. Generally, `DomainStructure` seems like a reasonable place to have the serialized volume data associated with an element.

Reply:
My suggestion: Move `TensorData.*pp` to `H5` (that's where it should be anyway I think). Then depend H5 on DomainStructure. Gives us the option of working with things like ElementId more directly in H5 volume data later, and DomainStructure will most likely never need to depend on H5. Thoughts?

Reply:
Yep, that sounds good to me!

Yea... The ReadSpec... should not be there in my opinion...